be/src/exec/olap_scan_node.cpp

```diff
@@ -397,8 +410,21 @@ StatusOr<pipeline::MorselQueuePtr> OlapScanNode::convert_scan_range_to_morsel_qu
         morsels.emplace_back(std::make_unique<pipeline::ScanMorsel>(node_id, scan_range));
     }
 
+    if (partition_order_hint().has_value()) {
+        bool asc = partition_order_hint().value();
+        std::stable_sort(morsels.begin(), morsels.end(), [asc](auto& l, auto& r) {
+            auto l_partition_id = down_cast<pipeline::ScanMorsel*>(l.get())->partition_id();
+            auto r_partition_id = down_cast<pipeline::ScanMorsel*>(r.get())->partition_id();
+            if (asc) {
+                return std::less()(l_partition_id, r_partition_id);
+            } else {
+                return std::greater()(l_partition_id, r_partition_id);
+            }
+        });
+    }
+
     if (output_chunk_by_bucket()) {
-        std::sort(morsels.begin(), morsels.end(), [](auto& l, auto& r) {
+        std::stable_sort(morsels.begin(), morsels.end(), [](auto& l, auto& r) {
             return down_cast<pipeline::ScanMorsel*>(l.get())->owner_id() <
                    down_cast<pipeline::ScanMorsel*>(r.get())->owner_id();
         });
```

Comment:
Code Duplication: The std::stable_sort() function is used twice in the last section of the code albeit with different conditions. Depending on the data and operations used, it might be more efficient to first sort by owner id and then proceed with partition id if that fits the requirements.

Reply:
it was not duplication code. partition_order_hint means we prefer order by morsel by partition_id, output_by_bucket means we must order by morsel by bucket_id. 
so we need a such result: order by bucket_id, partition_id. stable_sort by partition_id, bucket_id will get a expected result

Reply:
comment added